core/src/main/java/org/lflang/analyses/dag/Dag.java

```diff
@@ -0,0 +1,319 @@
+        label = "label=\"Dummy" + "=" + node.timeStep;
+        auxiliaryNodes.add(i);
+      } else if (node.nodeType == DagNode.dagNodeType.REACTION) {
+        label =
+            "label=\""
+                + node.nodeReaction.getFullName()
+                + "\n"
+                + "WCET="
+                + node.nodeReaction.wcet
+                + (node.getWorker() >= 0 ? "\nWorker=" + node.getWorker() : "");
+      } else {
+        // Raise exception.
+        System.out.println("UNREACHABLE");
+        System.exit(1);
+      }
+
+      // Add debug message, if any.
+      label += node.getDotDebugMsg().equals("") ? "" : "\n" + node.getDotDebugMsg();
+      // Add fillcolor and style
+      label += "\", fillcolor=\"" + node.getColor() + "\", style=\"filled\"";
+
+      code += i + "[" + label + "]";
+      dot.pr(code);
+    }
+
+    // Align auxiliary nodes.
+    dot.pr("{");
+    dot.indent();
+    dot.pr("rank = same;");
+    for (Integer i : auxiliaryNodes) {
+      dot.pr(i + "; ");
+    }
+    dot.unindent();
+    dot.pr("}");
+
+    // Add edges
+    for (DagNode source : this.dagEdges.keySet()) {
+      HashMap<DagNode, DagEdge> inner = this.dagEdges.get(source);
+      if (inner != null) {
+        for (DagNode sink : inner.keySet()) {
+          int sourceIdx = dagNodes.indexOf(source);
+          int sinkIdx = dagNodes.indexOf(sink);
+          dot.pr(sourceIdx + " -> " + sinkIdx);
+        }
+      }
+    }
+
+    dot.unindent();
+    dot.pr("}");
+
+    return this.dot;
+  }
+
+  public void generateDotFile(Path filepath) {
+    try {
+      CodeBuilder dot = generateDot();
+      String filename = filepath.toString();
+      dot.writeToFile(filename);
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+
+  /**
+   * Parses the dot file, reads the edges and updates the DAG. We assume that the edges are
+   * specified as: <SrcNodeId> -> <SinkNodeId>.
+   *
+   * <p>Furthermore, we assume that the new DAG (contained in the dotFile) will not have unnecessary
+   * edges, since they are removed by the shceduler.
+   *
+   * @param dotFilename
+   * @return
+   */
+  public boolean updateDag(String dotFileName) throws IOException {
+    FileReader fileReader;
+    BufferedReader bufferedReader;
+    // Read the file
+    try {
+      fileReader = new FileReader(dotFileName);
+      // Buffer the input stream from the file
+      bufferedReader = new BufferedReader(fileReader);
+    } catch (IOException e) {
+      System.out.println("Problem accessing file " + dotFileName + "! " + e);
+      return false;
+    }
+
+    String line;
+
+    // Pattern with which an edge starts:
+    Pattern pattern = Pattern.compile("^((\s*)(\\d+)(\s*)->(\s*)(\\d+))");
+    Matcher matcher;
+
+    // Before iterating to search for the edges, we clear the DAG edges array list
+    this.dagEdges.clear();
+
+    // Search
+    int i = 0;
+    while (bufferedReader.ready()) {
+
+      line = bufferedReader.readLine();
```

Comment:
Looping over a buffered reader should not be done by polling `ready`.
```suggestion
    while ((line = bufferedReader.readLine()) != null) {
```